apis/ingress/v1alpha1/astartedefaultingress_webhook.go

```diff
@@ -0,0 +1,248 @@
+
+	if theConfig, err = config.GetConfig(); err != nil {
+		return err
+	}
+
+	if c, err = client.New(theConfig, client.Options{}); err != nil {
+		return err
+	}
+
+	var allErrors field.ErrorList
+	var astarteFoundErr *field.Error
+	var astarte *apiv1alpha1.Astarte
+
+	if astarte, astarteFoundErr = r.validateReferencedAstarte(c); astarteFoundErr != nil {
+		allErrors = append(allErrors, astarteFoundErr)
+	}
+	// if astarte is not found, do not check the broker config as it is not available anyway
+	if astarteFoundErr == nil {
+		if err := r.validateBrokerTLSConfig(astarte); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	if err := r.validateBrokerServiceType(); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if err := r.validateAPITLSConfig(); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if err := r.validateDashboardTLSConfig(); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if err := r.validateRootTLSSecretExistence(c); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if err := r.validateAPITLSSecretExistence(c); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if err := r.validateDashboardTLSSecretExistence(c); err != nil {
+		allErrors = append(allErrors, err)
+	}
+	if len(allErrors) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "ingress", Kind: "AstarteDefaultIngress"},
+		r.Name,
+		allErrors,
+	)
+}
+
+func (r *AstarteDefaultIngress) validateBrokerServiceType() *field.Error {
+	if pointy.BoolValue(r.Spec.Broker.Deploy, true) {
+		if r.Spec.Broker.ServiceType == v1.ServiceTypeNodePort || r.Spec.Broker.ServiceType == v1.ServiceTypeLoadBalancer {
+			return nil
+		}
+
+		fldPath := field.NewPath("spec").Child("broker").Child("serviceType")
+		err := errors.New("Wrong broker service type. Allowed values: LoadBalancer, NodePort.")
+
+		astartedefaultingresslog.Error(err, "Allowed service types for the Broker are: LoadBalancer and NodePort")
+		return field.Invalid(fldPath, r.Spec.Broker.ServiceType, err.Error())
+	}
+	return nil
+}
+
+func (r *AstarteDefaultIngress) validateReferencedAstarte(c client.Client) (*apiv1alpha1.Astarte, *field.Error) {
+	fldPath := field.NewPath("spec").Child("astarte")
+
+	// ensure that the referenced Astarte instance exists
+	theAstarte := &apiv1alpha1.Astarte{}
+	if err := c.Get(context.Background(), types.NamespacedName{Name: r.Spec.Astarte, Namespace: r.Namespace}, theAstarte); err != nil {
+		astartedefaultingresslog.Error(err, "The referenced Astarte does not exist.")
+		return nil, field.Invalid(fldPath, r.Spec.Astarte, err.Error())
+	}
+	return theAstarte, nil
+}
+
+func (r *AstarteDefaultIngress) validateBrokerTLSConfig(astarte *apiv1alpha1.Astarte) *field.Error {
+	if pointy.BoolValue(astarte.Spec.VerneMQ.SSLListener, false) && astarte.Spec.VerneMQ.SSLListenerCertSecretName == "" {
+		err := errors.New("Broker TLS is misconfigured. Review your Astarte CR to ensure TLS termination at VerneMQ level.")
```

Comment:
I am wondering - is this where we validate it? I expect this check should be in the Astarte resource, it makes a lot more sense. Also, VMQ container would trigger a tons of errors, so this check definitely should not belong here but rather in the Astarte resource webhook.

Reply:
I think the validation that has to happen at this level is of a different kind. *IF* the Broker ingress is supposed to be deployed, *THEN* the VerneMQ SSL Listener *MUST* have been enabled in the Astarte resource. Instead, you're validating something that indeed you should take for granted from the Astarte resource, but you're letting this work in case the SSL listener wasn't enabled and you still requested the Broker ingress to be deployed.